### PR TITLE
configured security

### PR DESCRIPTION
**Pull request by Nazar090**

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
 import org.springframework.security.config.Customizer;
 import org.springframework.security.config.annotation.web.builders.HttpSecurity;
 import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
 import org.springframework.security.config.http.SessionCreationPolicy;
-import org.springframework.security.core.userdetails.UserDetailsService;
 import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
 import org.springframework.security.crypto.password.PasswordEncoder;
 import org.springframework.security.web.SecurityFilterChain;
 
+@Configuration
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/config/SecurityConfig.java:15]:
```suggestion
@Configuration
@RequiredArgsConstructor
```

File: src/main/java/com/example/onlinebookstore/repository/UserRepository.java [hunk 2]
```diff
@@ -3,9 +3,12 @@
...
 import java.util.Optional;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
 
 public interface UserRepository extends JpaRepository<User, Long> {
     boolean existsByEmail(String email);
 
-    Optional<User> findByEmail(String email);
+    @Query("SELECT u FROM User u JOIN FETCH u.roles WHERE u.email = :email")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/repository/UserRepository.java]: Use EntityGraph annotation to increase readability

File: src/main/java/com/example/onlinebookstore/security/CustomUserDetailsService.java [hunk 3]
```diff
@@ -5,17 +5,20 @@
 import org.springframework.security.core.userdetails.UserDetails;
 import org.springframework.security.core.userdetails.UserDetailsService;
 import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Component;
 import org.springframework.stereotype.Service;
 
 @Service
+@Component
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/security/CustomUserDetailsService.java]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/security/CustomUserDetailsService.java [hunk 4]
```diff
@@ -5,17 +5,20 @@
...
+public class CustomUserDetailsService implements UserDetailsService {
     private final UserRepository userRepository;
 
     @Override
     public UserDetails loadUserByUsername(String email)
             throws UsernameNotFoundException {
-        return userRepository.findByEmail(email)
-                .orElseThrow(() -> new UsernameNotFoundException("Can't find user by email"));
+        return userRepository.findByEmailWithRoles(email)
+                .orElseThrow(() -> new UsernameNotFoundException(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/security/CustomUserDetailsService.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException(
```

File: src/main/java/com/example/onlinebookstore/service/UserServiceImpl.java [hunk 5]
```diff
@@ -2,26 +2,44 @@
...
+        user.setRoles(Set.of(roleRepository.findByRole(Role.RoleName.USER)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find role by role name"
+                                + Role.RoleName.USER.name()))));
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
+        if (requestDto.getShippingAddress() != null) {
+            user.setShippingAddress(requestDto.getShippingAddress());
+        }
         userRepository.save(user);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/UserServiceImpl.java]:
```suggestion
        User user = userMapper.toUserEntity(requestDto);
        roleRepository.findByRole(Role.RoleName.USER)......
        user.setRoles(Set.of(role));
        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
        userRepository.save(user);
        return 
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/resources/db/changelog/changes/02-create-roles-table.yaml
Comment (sarakhmen) [inline src/main/resources/db/changelog/changes/02-create-roles-table.yaml:1]: why did you remove this file?

**Review by sarakhmen — APPROVED** [review]

File: src/main/resources/db/changelog/changes/02-create-roles-table.yaml
Comment (Nazar090) [reply src/main/resources/db/changelog/changes/02-create-roles-table.yaml:1]: i created a new one instead of the previous:) my bad
